### DISCUSSION TITLE
Mirrors, arbitrary shape, non-paraxial

### DISCUSSION THREAD

**Discussion by vitalychepel** [mjhoptics/ray-optics#140] [General]
Dear Michael,

I am interested in reconstruction of images of a series of point source by a spherical mirror (further - by a more complex surface with rotational symmetry defined, say, by a spline), which is set inside a glass cylinder, all viewed through a camera lens. I was looking to your python rayoptics page (https://ray-optics.readthedocs.io/en/latest/index.html#) but could not find any mentioning of reflecting surfaces. Can the package handle them? Another detail - the ray traces are expected to be far from the paraxial conditions. Still manageable?

Appreciate your advice. 
Best,
Vitaly Chepel

**Comment by mjhoptics** [comment]
Hello @vitalychepel,

Thank you for the question. Yes, **ray-optics** can handle reflecting surfaces, of potentially arbitrary profile. The ray-tracing is based on Snell's Law, without the paraxial approximation. As long as geometrical optics is a satisfactory approximation, **ray-optics** ray tracing should be fine.

I have a separate repo of jupyter notebooks that demonstrate different capabilities. Two notebooks in particular are worth noting:

[Off-axis parabola notebook v3.0](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/Off-axis%20parabola%20notebook%20v3.0.ipynb) Demonstrates spot diagram and geometrical PSF with a parabolic reflector.
[focus tilt notebook](https://github.com/mjhoptics/ray-optics-notebooks/blob/master/focus%20tilt%20notebook.ipynb) Also uses a parabolic reflector but demonstrates wavefront analysis and diffraction PSF

**ray-optics** doesn't have an implementation of a spline surface currently. **ray-optics** does support:

- [Conic](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.Conic)
- [Even (radial) Polynomial](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.EvenPolynomial)
- [Radial Polynomial](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.RadialPolynomial)
- [X and Y Aspheric Toroids](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.YToroid)

These are all part of the [profiles](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#module-rayoptics.elem.profiles) module. To add a new surface profile, one would write a class that inherits from [SurfaceProfile](https://ray-optics.readthedocs.io/en/stable/api/rayoptics.elem.profiles.html#rayoptics.elem.profiles.SurfaceProfile) and implements functions that evaluate the surface profile function and gradient at a point, as well as a few other support functions. Given this information, the **ray-optics** ray trace can calculate ray/surface intersections and refraction/reflection in a general way.

Please feel free to ask further questions. Thanks!
MIke Hayford
